scripts/usb_scan.py

```diff
@@ -0,0 +1,660 @@
+        try:
+            with open(self._PATH, "r") as f:
+                log_wrapper("=== policy file begin")
+                for line in f:
+                    log_wrapper(line[0:-1])
+                    self.parse_line(line)
+                log_wrapper("=== policy file end")
+        except IOError:
+            # without policy file, no device will be allowed to passed through
+            log_wrapper(traceback.format_exc())
+
+        log_wrapper("=== rule list begin")
+        log_list_wrapper(self.rule_list)
+        log_wrapper("=== rule list end")
+
+    def check_hex_length(self, name, value):
+        if name in [self._CLASS, self._SUBCLASS, self._PROTOCOL]:
+            return 2 == len(value)
+        if name in [self._ID_VENDOR, self._ID_PRODUCT, self._BCD_DEVICE]:
+            return 4 == len(value)
+        return False
+
+    @staticmethod
+    def log_parse_error(pos, end, target, line):
+        log.error(
+            "Malformed policy rule, unable to parse '{}'. Ignored line: {}"
+            .format(target[pos:end], line))
+
+    def parse_line(self, line):
+        """ parse one line of policy file, generate rule, and append it to
+        self.rule_list
+
+        Example:
+        ALLOW:vid=056a pid=0315 class=03 # Wacom Intuos tablet
+            => {'pid': '0315', 'vid': '056a', 'allow': True, 'class': '03'}
+        DENY: class=03 subclass=01 prot=01 # HID Boot keyboards
+            => {'prot': '01', 'class': '03', 'allow': False, 'subclass': '01'}
+        ALLOW: # Otherwise allow everything else
+            => {'allow': True}
+
+        :param line: (str) single line of policy file
+        :return: None
+        """
+        # 1. remove comments
+        # ^([^#]*)(#.*)?$
+        i = line.find("#")
+        if i > 0:
+            line = line[0:i]
+        elif i == 0:
+            return
+
+        # 2. split action and match field
+        # ^\s*(ALLOW|DENY)\s*:\s*([^:]*)$
+        try:
+            (action, target) = [part.strip() for part in line.split(":")]
+        except ValueError as e:
+            if line.rstrip():
+                log.error(
+                    "Caught error {}. Ignoring malformed rule line: {}"
+                    .format(str(e), line))
+            return
+
+        # 3. parse action
+        # \s*(ALLOW|DENY)\s*
+        rule = {}
+        if action.lower() == "allow":
+            rule[self._ALLOW] = True
+        elif action.lower() == "deny":
+            rule[self._ALLOW] = False
+        else:
+            log.error("Malformed action'{}', ignoring rule line: {}".format(
+                action, line))
+            return
+
+        # 4. parse key=value pairs
+        # pattern = r"\s*(class|subclass|prot|vid|pid|rel)\s*=\s*([0-9a-f]+)"
+        last_end = 0
+        for matchNum, match in enumerate(re.finditer(self._PATTERN, target,
+                                                     re.IGNORECASE)):
+            tmp_end = last_end
+            last_end = match.end()
```

Comment:
I think that you can move this assignement before the try block, and avoid having the `tmp_end` without messing upa with the readability.

Reply:
I want to check the updated last_end if return from the first if cause below.

Reply:
Yes I understand. I was suggesting to do:

```python
if last_end != match.start():
                self.log_parse_error(last_end, match.start(), target, line)
                return
if len(match.groups()) != 2 or not all(match.groups()):
                self.log_parse_error(match.start(), match.end(), target, line)
                return
last_end = match.end()
```

Reply:
I think `len(match.groups())` cannot be different from 2 as you regex has two groups.
Also this last check is making the next try block useless.

Let's simplify and rewrite the whole block as:
```python
        last_end = 0
        for matchNum, match in enumerate(re.finditer(self._PATTERN, target,
                                                     re.IGNORECASE)):
            if tmp_end != match.start():
                self.log_parse_error(tmp_end, match.start(), target, line)
                return
            if len(match.groups()) != 2 or not all(match.groups()):
                self.log_parse_error(match.start(), match.end(), target, line)
                return
            last_end = match.end()

            try:
                (name, value) = [part.lower() for part in match.groups()]
                if not self.check_hex_length(name, value):
                    log.error("hex'{}' length error, ignore line {}".format(
                              str(value), line))
                    return
                if name not in self.rule_list:
                    rule[name] = value
            except AttributeError:
                self.log_parse_error(match.start(), match.end(), target, line)
                return
            except ValueError:
                self.log_parse_error(match.start(), match.end(), target, line)
                return
```

Reply:
done.